bigquery/storage/apiv1beta2/thickwriter.go

```diff
@@ -0,0 +1,244 @@
+			// TODO: how do we signal stopped?
+			break
+		}
+		if err != nil {
+			// observed errors here
+			// open question: should these be associated to the next AR, the writer, or both?
+
+			// when you don't send schema:
+			// code = InvalidArgument desc = Invalid proto schema: BqMessage.proto: : Missing name. Entity: projects/shollyman-demo-test/datasets/storage_test_dataset_20210122_70895612072539_0002/tables/testtable_20210122_70895612106658_0002/_default
+			// slow metadata update?
+			// code = PermissionDenied desc = Permission 'TABLES_UPDATE_DATA' denied on resource 'projects/shollyman-demo-test/datasets/storage_test_dataset_20210122_70746896961953_0002/tables/testtable_20210122_70746897001459_0002' (or it may not exist).
+			log.Printf("\ngot an error from recv: %v\n", err)
+
+			// for now, kill the loop until we figure out
+			break
+		}
+
+		// take the lock while we process an element from the queue
+		tw.mu.Lock()
+
+		// get pending AR from queue.
+		nextAR := tw.pending[0]
+		tw.pending = tw.pending[1:]
+
+		if status := resp.GetError(); status != nil {
+			// TODO: add retry logic here.
+			// need a custom status error?
+			nextAR.err = fmt.Errorf("%d: %s", status.GetCode(), status.GetMessage())
+		}
+
+		if success := resp.GetAppendResult(); success != nil {
+			if off := success.GetOffset(); off != nil {
+				nextAR.offset = off.GetValue()
+			}
+		}
+
+		// clear the request (may never need this, once we resolve the retry strategy)
+		nextAR.req = nil
+		// signal this write complete
+		close(nextAR.ready)
+		// done for now, leave the lock
+		tw.mu.Unlock()
+	}
+}
+
+type AppendResult struct {
+	req    *storagepb.AppendRowsRequest
+	ready  chan struct{}
+	err    error
+	offset int64
+}
+
+func (ar *AppendResult) Ready() <-chan struct{} { return ar.ready }
+
+func (ar *AppendResult) GetResult(ctx context.Context) (int64, error) {
+	select {
+	case <-ar.Ready():
+		return ar.offset, ar.err
+	default:
+	}
+
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case <-ar.Ready():
+		return ar.offset, ar.err
+	}
+}
+
+func NewAppendResult() *AppendResult {
+	return &AppendResult{ready: make(chan struct{})}
+}
+
+func (tw *ThickWriter) Stop(ctx context.Context) error {
+	// need to close down processing of stream responses
+	return fmt.Errorf("unimplemented")
+}
+
+func (tw *ThickWriter) RegisterProto(in proto.Message) error {
+	if in == nil {
+		return fmt.Errorf("no proto provided")
+	}
+	tw.mu.Lock()
+	defer tw.mu.Unlock()
+	if tw.protoMessageDescriptor != nil {
+		return fmt.Errorf("already have a type registered")
+	}
+	tw.protoMessageDescriptor = protodesc.ToDescriptorProto(in.ProtoReflect().Descriptor())
+	return nil
+}
+
+// AppendRows expects a slice of proto.Messages, which it serialized and then appends.
+// It returns an AppendResult, which can be used to check that the write completed.
+func (tw *ThickWriter) AppendRows(protoSlice interface{}) (*AppendResult, error) {
```

Comment:
Is there anything we need to do to indicate this is thread-safe?

Reply:
No annotations, but doc wise we'll want to be clear.  Generally its a matter of identifying things that aren't thread safe.